### PR TITLE
convert to immutable for CUDA tests

### PR DESCRIPTION
**Pull request by m-bossart**
Converts NonlinearProblem to ImmutableNonlinearProblem for CUDA compatibility (SciML/NonlinearSolve.jl#439)

### CODE REVIEW

**Comment by m-bossart** [comment]
Tests are failing due to this methoderror. I'm not sure why given the constructor in the new `ImmutableNonlinearProblem `struct 
``
 MethodError: no method matching SimpleNonlinearSolve.ImmutableNonlinearProblem(::SciMLBase.NonlinearFunction{false, SciMLBase.FullSpecialize, typeof(Main.var"##CUDA Kernel Launch Test#1070".f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED_NO_TIME), Nothing, Nothing, Nothing}, ::StaticArraysCore.SVector{2, Float32}, ::SciMLBase.NullParameters, ::SciMLBase.StandardNonlinearProblem)
``

File: src/immutable_nonlinear_problem.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+end
+
+staticarray_itize(x) = x
+staticarray_itize(x::Vector) = SVector{length(x)}(x)
+staticarray_itize(x::SizedVector) = SVector{length(x)}(x)
+staticarray_itize(x::Matrix) = SMatrix{size(x)...}(x)
+staticarray_itize(x::SizedMatrix) = SMatrix{size(x)...}(x)
+
+function Base.convert(::Type{ImmutableNonlinearProblem}, prob::T) where {T <: NonlinearProblem}
+    ImmutableNonlinearProblem(prob.f,
```
Comment (ChrisRackauckas) [inline src/immutable_nonlinear_problem.jl:63]:
```suggestion
    ImmutableNonlinearProblem{isinplace(prob)}(prob.f,
```

File: src/SimpleNonlinearSolve.jl [hunk 2]
```diff
@@ -81,7 +82,18 @@ function SciMLBase.solve(prob::NonlinearProblem, alg::AbstractSimpleNonlinearSol
...
+function SciMLBase.solve(prob::Union{ImmutableNonlinearProblem}, alg::AbstractSimpleNonlinearSolveAlgorithm,
+    args...; sensealg = nothing, u0 = nothing, p = nothing, kwargs...)
+if sensealg === nothing && haskey(prob.kwargs, :sensealg)
+    sensealg = prob.kwargs[:sensealg]
+end
+new_u0 = u0 !== nothing ? u0 : prob.u0
+new_p = p !== nothing ? p : prob.p
+return __internal_solve_up(prob, sensealg, new_u0, u0 === nothing, new_p,
+    p === nothing, alg, args...; prob.kwargs..., kwargs...)
+end
```
Comment (ChrisRackauckas) [inline src/SimpleNonlinearSolve.jl:94]: formatting

File: src/SimpleNonlinearSolve.jl [hunk 3]
```diff
@@ -81,7 +82,18 @@ function SciMLBase.solve(prob::NonlinearProblem, alg::AbstractSimpleNonlinearSol
         p === nothing, alg, args...; prob.kwargs..., kwargs...)
 end
 
-function __internal_solve_up(_prob::NonlinearProblem, sensealg, u0, u0_changed,
+function SciMLBase.solve(prob::Union{ImmutableNonlinearProblem}, alg::AbstractSimpleNonlinearSolveAlgorithm,
+    args...; sensealg = nothing, u0 = nothing, p = nothing, kwargs...)
```
Comment (ChrisRackauckas) [inline src/SimpleNonlinearSolve.jl:86]: The conversion should go here.

Comment (m-bossart) [reply src/SimpleNonlinearSolve.jl]: I moved the conversion to inside the solve in the last commit. Now I'm getting a method error when precompiling due to the startup workload. It is complaining about this line when solving a problem that has been converted and its not clear to me how to fix it: https://github.com/SciML/SimpleNonlinearSolve.jl/blob/2e609d9fc952881f124e21bb8181a3a5c4e3b91a/src/nlsolve/lbroyden.jl#L185

**Comment by avik-pal** [comment]
I will review it in detail later, but tagging @Vaibhavdixit02 because it will very likely break GPUPSO. Maybe we tag a breaking release just for safety

**Comment by Vaibhavdixit02** [comment]
Since that package wasn't released, you should be fine on that end.

**Comment by ChrisRackauckas** [comment]
GPUPSO should be fine if this is done right? We just need to make sure the kernels use everything immutable.

File: test/gpu/cuda_tests.jl [hunk 4]
```diff
@@ -51,6 +51,7 @@ end
     end
 
     prob = NonlinearProblem{false}(f, @SVector[1.0f0, 1.0f0])
+    prob = convert(SimpleNonlinearSolve.ImmutableNonlinearProblem, prob)
```
Comment (ChrisRackauckas) [inline test/gpu/cuda_tests.jl:54]:
```suggestion
```

No cheating.

File: src/immutable_nonlinear_problem.jl [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+staticarray_itize(x) = x
+staticarray_itize(x::Vector) = SVector{length(x)}(x)
+staticarray_itize(x::SizedVector) = SVector{length(x)}(x)
+staticarray_itize(x::Matrix) = SMatrix{size(x)...}(x)
+staticarray_itize(x::SizedMatrix) = SMatrix{size(x)...}(x)
+
+function Base.convert(::Type{ImmutableNonlinearProblem}, prob::T) where {T <: NonlinearProblem}
+    ImmutableNonlinearProblem{isinplace(prob)}(prob.f,
+        staticarray_itize(prob.u0),
+        staticarray_itize(prob.p),
```
Comment (ChrisRackauckas) [inline src/immutable_nonlinear_problem.jl:65]: This isn't going to be type stable, we shouldn't copy this part. Don't automatically make things static array, just immutable the problem.

File: src/nlsolve/broyden.jl [hunk 6]
```diff
@@ -22,7 +22,7 @@ end
 
 __get_linesearch(::SimpleBroyden{LS}) where {LS} = Val(LS)
 
-function SciMLBase.__solve(prob::NonlinearProblem, alg::SimpleBroyden, args...;
+function SciMLBase.__solve(prob::Union{NonlinearProblem, ImmutableNonlinearProblem}, alg::SimpleBroyden, args...;
```
Comment (ChrisRackauckas) [inline src/nlsolve/broyden.jl:25]: None of these unions should be required. If done correctly, the top level should change to an ImmutableNonlinearProblem, and then they should all be that.

**Comment by ChrisRackauckas** [comment]
Left a few comments. If those two things are done this should be quick to finish.

**Comment by m-bossart** [comment]
I made changes according to your comments. The simple adjoint test is failing with this error:
`  Cannot determine ordering of Dual tags Nothing and ForwardDiff.Tag{Base.Fix2{SciMLBase.NonlinearFunction{false, SciMLBase.FullSpecialize, typeof(Main.var"##Simple Adjoint Test#264".ff), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED_NO_TIME), Nothing, Nothing, Nothing}, Vector{Float64}}, Float64}` 

Any thoughts on the fix for that?

**Comment by m-bossart** [comment]
These tests are also failing: https://github.com/SciML/SimpleNonlinearSolve.jl/blob/40958d2c2fdff0986d113ea1f5ed1ffa96316e89/test/gpu/cuda_tests.jl#L40C1-L40C107

Doesn't the `kernel_function` need to change given that the conversion to an Immutable struct happens within the solve call? I'm not sure how that test should be restructured.

**Comment by ChrisRackauckas** [comment]
They kernel function needs to be the immutable form. And this would need an FAQ entry.

> Any thoughts on the fix for that?

Add the immutable stuff to the AD dispatches?

**Pull request closed without merging**